### PR TITLE
Hnswlib implementation

### PR DESCRIPTION
**Pull request by AmirYalamov**
## Description

We implemented the [hnswlib library](https://github.com/nmslib/hnswlib), which uses a Small World search method to find the nearest neighbour. The unit test for it works, and `main_describeAndMatch.cpp` also works using HNSWLIB as the algorithm.

## Features list

- [X] hnswlib library in the `/dependencies/` directory
- [X] `ArrayMatcher` implementation file for hnswlib 
- [X] Unit test to see if hnswlib methods work

## Implementation remarks

For the `ArrayMatcher_hnswlib.cpp` implementation, we followed the convention and style for the other `ArrayMatcher` files in the `/matching/` directory. We also followed the same convention for our unit test `hnsw_test.cpp`. We were wondering if we should keep the unit test for hnswlib separate, or add the testing in `matching_testing.cpp`?

1. In the new line 73 and 74, we commented out the code as show below. Is this particular code needed and should we remove the commenting? 
```    
// HNSWLIB returns true binary but our matcher type is HNSWLIB
//if (regions.IsBinary() && matcherType != BRUTE_FORCE_HAMMING)
//	return out;
```
2. For Visual Studio IDE: how do we generate a project so we can have the unit tests to run? Do we need to edit the CMake file?
3. To import the hnswlib library, we downloaded it and put it in the `/dependencies/` directory. Is this how you wanted us to do it? Is there a more elegant way to do this or is this way fine?

@flosincapite @simogasp @fabiencastan please review :)

### CODE REVIEW

**Comment by fabiencastan** [comment]
1/ Yes, I think you can remove it
2/ You should already have a target in VS with the name `alicevision_test_matching_hnsw` in the group "Test".
3/ I would prefer to use git submodule to https://github.com/nmslib/hnswlib.

Next step is to integrate it into:
https://github.com/alicevision/AliceVision/blob/develop/src/software/pipeline/main_featureMatching.cpp

And finally declare the new option on the Meshroom side:
https://github.com/alicevision/meshroom/blob/develop/meshroom/nodes/aliceVision/FeatureMatching.py

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
+// This file is part of the AliceVision project.
```
Comment (rjanvier) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:1]: Please keep the original copyright header:
https://github.com/openMVG/openMVG/blob/cf5a5754926b7e73c9e1d0f67fac296db2365e5f/src/openMVG/matching/matcher_hnsw.hpp#L1-L7. You didn't change the copyright header of HNSNlib, right?. IT is not a a fair open-source citizenship...

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        else
+        {
+            std::cerr << "HNSW matcher: this type of distance is not handled Yet" << std::endl;
+            return false;
+        }
+
+        HNSWmatcher.reset(new HierarchicalNSW<DistanceType>(HNSWmetric.get(), nbRows, 16, 100));
+        HNSWmatcher->setEf(16);
+
+        // add first point..
```
Comment (rjanvier) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:93]: Why did you trim OpenMP? I think you'll experiencing performance loss.

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        if(HNSWmatcher.get() == nullptr)
+        {
+            return false;
+        }
+
+        pvec_distances->reserve(nbQuery * NN);
+        pvec_indices->reserve(nbQuery * NN);
+
+		for(int i = 0; i < nbQuery; i++)
+        {
```
Comment (rjanvier) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:143]: same comment about openMP here

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 1 above)
Comment (fabiencastan) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:1]: Please keep a welcoming environment. This is a PR from students who are making their _first open source contribution_ without any bad intention.

For the header, as you and Pierre have been the original authors, then it has been integrated and maintained in OpenMVG, then integrated and maintained in AliceVision, I would suggest:
```
// This file is part of the AliceVision project.
// Copyright (c) 2020 AliceVision contributors.
// Copyright (c) 2019 openMVG contributors.
// Copyright (c) 2019 Romain Janvier and Pierre Moulon
// This Source Code Form is subject to the terms of the Mozilla Public License,
// v. 2.0. If a copy of the MPL was not distributed with this file,
// You can obtain one at https://mozilla.org/MPL/2.0/.
```
@pmoulon @rjanvier Is that fine to you?

Comment (rjanvier) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:1]: fine, thanks.

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 2 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:93]: Thank you for pointing this out. We did not use it because we followed the same format for the `#include` statements and general file layout as the other `ArrayMatcher` files in Alice Vision for matching. We will add OpenMP, test it, and then push the changes for you to see here.

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 3 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:143]: Same reply as above :)

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 1 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:1]: Apologies for this, we did not notice it; thank you for pointing it @rjanvier :). We will add the copyright header @fabiencastan suggested above.

File: src/aliceVision/matching/RegionsMatcher.hpp [hunk 4]
```diff
@@ -145,7 +145,7 @@ class RegionsMatcher : public IRegionsMatcher
     if (!matcher_.SearchNeighbours(queries, queryregions_.RegionCount(), &vec_nIndice, &vec_fDistance, NNN__))
       return false;
 
-    assert(vec_nIndice.size() == vec_fDistance.size());
+    //assert(vec_nIndice.size() == vec_fDistance.size());
```
Comment (AmirYalamov) [inline src/aliceVision/matching/RegionsMatcher.hpp:148]: So once I added OpenMP, this line would fail, but when I commented it out the sample code worked fine as before. Is this required, shouldn't we only be doing assertions in the unit test?

Comment (honsontran) [reply src/aliceVision/matching/RegionsMatcher.hpp:148]: @rjanvier , could it be possible the assertion is failing due to a race condition?

Comment (rjanvier) [reply src/aliceVision/matching/RegionsMatcher.hpp:148]: Sure it is :). you are missing an "OpenMP critical section" before the result loop [see here](https://github.com/openMVG/openMVG/blob/cf5a5754926b7e73c9e1d0f67fac296db2365e5f/src/openMVG/matching/matcher_hnsw.hpp#L147-L149)

this make me think you can also try direct indexing, in theory it's far better, but in practice I bet you won't notice the difference. Something like that:
```
  pvec_indices->resize(nbQuery * NN);
  pvec_distances->resize(nbQuery * NN);
...
...
    #Your current OpenMP pragma
    for (int i = 0; i < nbQuery; ++i) {
      const auto result = HNSWmatcher->searchKnn((const void *) (query + dimension_ * i), NN,
        [](const std::pair<DistanceType, size_t> &a, const std::pair<DistanceType, size_t> &b) -> bool {
          return a.first < b.first;
      });
      for (int j = 0; j < NN; ++j)
      {
        const auto & res = result[j];
        (*pvec_indices)[i*NN+j] = IndMatch(i, res.second); ;
        (*pvec_distances)[i*NN+j] = res.first;
      }
    }```

Comment (AmirYalamov) [reply src/aliceVision/matching/RegionsMatcher.hpp]: Ah my bad, silly mistake to forget that critical section. I have added it, and I removed the commenting out of that one assertion statement in `RegionsMatcher.hpp`, and it doesn't fail anymore; everythings works great!

File: src/aliceVision/matching/RegionsMatcher.cpp [hunk 5]
```diff
@@ -10,172 +10,190 @@
 #include "aliceVision/matching/ArrayMatcher_bruteForce.hpp"
 #include "aliceVision/matching/ArrayMatcher_kdtreeFlann.hpp"
 #include "aliceVision/matching/ArrayMatcher_cascadeHashing.hpp"
+#include "aliceVision/matching/ArrayMatcher_hnswlib.hpp"
```
Comment (simogasp) [inline src/aliceVision/matching/RegionsMatcher.cpp:14]: Please reformat le code using spaces instead of tabs

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+#include <vector>
+
+#include <random>
+#include <cmath>
+
+using hnswlib::HierarchicalNSW;
+using hnswlib::L2Space;
+using hnswlib::L2SpaceI;
+using hnswlib::SpaceInterface;
+using namespace std;
```
[the remark this answers is not in the document]
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:34]: Just to clarify you're saying to remove only `using namespace std;` and not line 31 to 33 that you quoted right?

[the remark this answers is not in the document]
Comment (simogasp) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:34]: yes remove L34.

And actually it would be better to move those other `using` inside the class, after `using  DistanceType = typename Metric::ResultType;` for example

File: src/aliceVision/CMakeLists.txt [hunk 7]
```diff
@@ -32,12 +32,16 @@ if(ALICEVISION_BUILD_SFM)
   add_subdirectory(voctree)
 
   if(ALICEVISION_HAVE_OPENCV)
-    add_subdirectory(calibration)
+    add_subdirectory(hdr)
```
Comment (fabiencastan) [inline src/aliceVision/CMakeLists.txt:35]: No reason to change that. Need to revert this change.

File: src/aliceVision/CMakeLists.txt [hunk 8]
```diff
@@ -32,12 +32,16 @@ if(ALICEVISION_BUILD_SFM)
...
   endif()
 endif()
 
 if(ALICEVISION_BUILD_HDR)
-add_subdirectory(hdr)
+  add_subdirectory(hdr)
+endif()
+
+if(ALICEVISION_BUILD_HDR OR ALICEVISION_BUILD_MVS)
+  add_subdirectory(mvsData)
```
Comment (fabiencastan) [inline src/aliceVision/CMakeLists.txt:44]: This creates the CI failure.
You need to remove the duplicated line (line 52), to avoid having the same `add_subdirectory` twice.
So keep this line, but remove the one inside `if(ALICEVISION_BUILD_MVS)`.

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:34]: For moving L31 - L33 inside the class after the line `using DistanceType = typename Metric::ResultType;` I am getting an error `Using declaration in class refers into 'hnswlib::', which is not a class`, which I think has to do with scope. I tried looking for answers online but I wasn't able to find anything. Do I need to each of the lines somehow to look more like the line before `using DistanceType = typename Metric::ResultType;`?

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+     * \param[in] nbRows    The number of component.
+     * \param[in] dimension Length of the data contained in the dataset.
+     *
+     * \return True if success.
+     */
+    bool Build(const Scalar* dataset, int nbRows, int dimension)
+    {
+        HNSWmetric.reset(nullptr);
+        HNSWmatcher.reset(nullptr);
+        dimension_ = 0;
```
[the remark this answers is not in the document]
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:73]: Would the correct approach be to do replace those lines with `reset();` and then add the function 
```
void reset() {
        HNSWmetric.reset(nullptr);
        HNSWmatcher.reset(nullptr);
        dimension_ = 0;
    }
```
in the `private;` section of `ArrayMatcher_hnswlib.hpp`?

[the remark this answers is not in the document]
Comment (simogasp) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:73]: Yes, personally i like it better

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+            HNSWmatcher->addPoint((void*)(dataset + dimension * i), (size_t)i);
+        }
+
+        return true;
+    };
+
+    /**
+     * Search the nearest Neighbor of the scalar array query.
+     *
+     * \param[in]   query     The query array
```
Comment (simogasp) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:115]: same here and in the other methods `@param` etc

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (simogasp) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp]: ah yes my bad, it's not a derived class. You can leave them like that.

File: src/aliceVision/CMakeLists.txt [hunk 11]
```diff
@@ -49,8 +49,7 @@ endif()
 if(ALICEVISION_BUILD_MVS)
   add_subdirectory(lightingEstimation)
   add_subdirectory(mesh)
-  add_subdirectory(mvsData)
-  add_subdirectory(mvsUtils)
+    add_subdirectory(mvsUtils)
```
Comment (simogasp) [inline src/aliceVision/CMakeLists.txt:52]: i think it should be:
```suggestion
    add_subdirectory(mvsData)
    add_subdirectory(mvsUtils)
```
https://github.com/alicevision/AliceVision/blob/develop/src/aliceVision/CMakeLists.txt#L48

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 12]
```diff
@@ -178,6 +176,12 @@ class ArrayMatcher_hnswlib : public ArrayMatcher<Scalar, Metric>
     int dimension_{0};
     std::unique_ptr<SpaceInterface<DistanceType>> HNSWmetric;
     std::unique_ptr<HierarchicalNSW<DistanceType>> HNSWmatcher;
+
+    void reset() {
+        HNSWmetric.reset(nullptr);
+        HNSWmatcher.reset(nullptr);
+        dimension_ = 0;
+    }
```
Comment (simogasp) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:184]: it's better to separate public/private methods from member:

```c++
private:
    void reset() 
    {
￼        HNSWmetric.reset(nullptr);
￼        HNSWmatcher.reset(nullptr);
￼        dimension_ = 0;
￼    }

private:
    int dimension_{0};
￼    std::unique_ptr<SpaceInterface<DistanceType>> HNSWmetric;		￼    
￼    std::unique_ptr<HierarchicalNSW<DistanceType>> HNSWmatcher;		￼    

```

File: src/aliceVision/CMakeLists.txt (hunk 11 above)
Comment (AmirYalamov) [reply src/aliceVision/CMakeLists.txt:52]: @fabiencastan suggested above 
> This creates the CI failure.
> You need to remove the duplicated line (line 52), to avoid having the same `add_subdirectory` twice.
> So keep this line, but remove the one inside `if(ALICEVISION_BUILD_MVS)`.

so to have the code as 
```
if(ALICEVISION_BUILD_HDR)
  add_subdirectory(hdr)
endif()

if(ALICEVISION_BUILD_HDR OR ALICEVISION_BUILD_MVS)
  add_subdirectory(mvsData)
endif()

# MVS modules

if(ALICEVISION_BUILD_MVS)
  add_subdirectory(lightingEstimation)
  add_subdirectory(mesh)
    add_subdirectory(mvsUtils)
  add_subdirectory(fuseCut)
```
I think the issue is using `mvsData` twice. Your suggestion is good, but that would mean we would have to remove 
```
if(ALICEVISION_BUILD_HDR OR ALICEVISION_BUILD_MVS)
  add_subdirectory(mvsData)
endif()
```
Would that be OK to do?

File: src/aliceVision/matching/RegionsMatcher.cpp (hunk 5 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/RegionsMatcher.cpp:14]: I just pushed the reformatting changes, please review and let me know if I missed a spot to reformat :)

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp (hunk 12 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/ArrayMatcher_hnswlib.hpp]: Just pushed this change!

File: src/aliceVision/matching/RegionsMatcher.hpp (hunk 4 above)
Comment (AmirYalamov) [reply src/aliceVision/matching/RegionsMatcher.hpp]: @rjanvier Awesome, thank you for the direct indexing solution, about to push it soon. I was just curious, other than avoiding local variable issues during multi-threading, what are the advantages of the direct indexing solution?

Comment (rjanvier) [reply src/aliceVision/matching/RegionsMatcher.hpp]: In theory, by using "critical section", only one thread can do the update at time so it's possible you loose performances somehow (because some threads could be waiting their turn to emplace in the vector)... 
But In theory too, `vector::resize` should have an overhead vs. `vector::reserve + emplace_back`, so it's a mixed bag and it depends also on how compilers optimize this. On my side, with a  micro becnhmark of 10 000k pairs of images, I noticed the two approaches are surprisingly very similar in term of execution speed... but I do prefer the semantic of "direct indexing" (and I think everybody should)

Comment (AmirYalamov) [reply src/aliceVision/matching/RegionsMatcher.hpp]: That makes sense, with very similar performance a static approach rather than multithreading seems safer. I also just pushed the direct indexing approach, please let me know if anything could be improved or changed.

Comment (rjanvier) [reply src/aliceVision/matching/RegionsMatcher.hpp]: I'm not a software engineer and I think you have more clever and skilled people than me at your disposal. I'm confident they can help you to sort that out :)

**Comment by AmirYalamov** [comment]
@fabiencastan  I have added the necessary options for hnswlib (I think, please let me know if I'm missing something or put something in the wrong place). 
> Next step is to integrate it into:
> https://github.com/alicevision/AliceVision/blob/develop/src/software/pipeline/main_featureMatching.cpp
> 
> And finally declare the new option on the Meshroom side:
> https://github.com/alicevision/meshroom/blob/develop/meshroom/nodes/aliceVision/FeatureMatching.py

Here is the [related Meshroom PR](https://github.com/alicevision/meshroom/pull/998).

Please let me know if there is anything else we need to take care of for this PR :) @flosincapite @simogasp @fabiencastan

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 13]
```diff
@@ -137,52 +137,52 @@ class ArrayMatcher_hnswlib : public ArrayMatcher<Scalar, Metric>
...
+        return false;
+      }
+
+      pvec_distances->resize(nbQuery * NN);
+      pvec_indices->resize(nbQuery * NN);
+
+      #ifdef ALICEVISION_IS_DEFINED
+      #pragma omp parallel for
+      #endif
+      for(int i = 0; i < nbQuery; i++)
```
Comment (simogasp) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:151]:
```suggestion
      for(int i = 0; i < nbQuery; ++i)
```
nitpicking but is alway better to prefer preincrement in these cases.

File: src/aliceVision/matching/ArrayMatcher_hnswlib.hpp [hunk 14]
```diff
@@ -137,52 +137,52 @@ class ArrayMatcher_hnswlib : public ArrayMatcher<Scalar, Metric>
...
+          [](const std::pair<DistanceType, size_t>& var1, const std::pair<DistanceType, size_t>& var2) -> bool {
+              return var1.first < var2.first;
+          });
+
+        /* direct indexing approach chosen over omp critical (OpenMP) approach due
+         * to possible performance loss from waiting threads in OpenMP approach
+        */
         {
-            return false;
+          for (int j = 0; j < NN; j++) {
```
Comment (simogasp) [inline src/aliceVision/matching/ArrayMatcher_hnswlib.hpp:163]:
```suggestion
          for (int j = 0; j < NN; ++j) {
```

File: src/aliceVision/CMakeLists.txt [hunk 15]
```diff
@@ -40,16 +40,13 @@ if(ALICEVISION_BUILD_HDR)
   add_subdirectory(hdr)
 endif()
```
Comment (fabiencastan) [inline src/aliceVision/CMakeLists.txt:42]: This file was correct before this change
